### ISSUE TITLE
Fix float/double mismatches in source

### ISSUE THREAD

**Issue by hollasch** [RayTracing/InOneWeekend#67]
Lots of places where we're using `double` constants for use in `float` calculations, mixing up our widths when passing parameters, and such. We need to ensure that our code compiles without any warnings about mixing precision.

Ideally, it would be good to determine what the impact of `float` vs. `double`. Personally, I prefer double precision, as it's the native precision for math, and C++ constants, and is frequently faster on platforms that perform intrinsic computation in double-precision (and thus induce a performance hit when transforming from and to single-precision.

**Comment by trevordblack** [comment]
A possible solution here is to replace all decimal precision types with a
"real" type that is typedef'd to doubles or floats

**Comment by trevordblack** [comment]
I don't believe that my comment above is warranted. We should pick a precision type and stick with it.

I went ahead and created a benchmark.
On my own 
Intel® Core™ i7-9750H CPU @ 2.60GHz × 12
running in Ubuntu 18.04.3 LTS
Compiled with Clang6.0.0 ubuntu
I found:
double: 442 seconds
float: 456 seconds
The difference is noise.

If more data is warranted,
$ git pull
$ git checkout origin/tdb/real_type_exp
$ cd src
$ mkdir build
$ cd build
$ cmake ..
$ make
$ time ./Real_Type_Double > double.ppm && time ./Real_Type_Float > float.ppm

**To Be Clear**
This is not a production branch.
Once a decision of floats vs doubles is made, I can go through and replace `Real`s with the selection.

**Comment by hollasch** [comment]
I also had the idea of `using real = double` or `using real = float`. However, your benchmark bears out my experience, that `float` is not a performance optimization, and indeed, working entirely in `double` values often yields a (tiny) performance benefit. Old habits die hard, though.

I say we convert the entire code base to use simple double-precision floats.

Anybody else care to comment?

**Comment by ronaldfw** [comment]
If double is not slower than float (single precision) here, I'd vote for just using double everywhere. With float you could get performance benefits when implementing certain parts with SIMD (SSE, AVX) but I don't think we should go there (and I don't know if there are even places where that would work well in any of the code here).

**Comment by hollasch** [comment]
I think we're starting to lean toward `double` at this point. The only reason to use single-precision is that GPUs typically work with floats, but we're not in that domain. Converting to float is fine for storage purposes, but we're not doing that either.

**Comment by trevordblack** [comment]
From all of this discussion, what I've arrived at is that the only significant difference between `float`s and `double`s for this domain is that `float`s are much more likely to have shadow acne.

This is arguably a benefit because it would much more clearly illucidate the problem.

Other than that, I'm not seeing any real argument that isn't aesthetic. My vote is for `double`s.

**Comment by hollasch** [comment]
I suspect that adopting double precision may _not_ change the frequency of surface acne. Does it matter if your intersection point is 1e-10 inside the surface versus 1e-100 inside? Should be a simple test to verify, but agree with Trevor that it shouldn't really affect the decision.

Sounds like we're agreed on using `double`.
